v8_c_api/src/v8_c_api.cpp

```diff
@@ -306,6 +304,356 @@ v8_pd_list* v8_PDListCreate(v8::ArrayBuffer::Allocator *alloc) {
+    void sendNotification(std::unique_ptr<v8_inspector::StringBuffer> message) override;
+    void flushProtocolNotifications() override;
+
+    void setIsolate(v8::Isolate *isolate);
+
+    void setOnResponseCallback(
+        const InspectorOnResponseCallback &callback,
+        const InspectorUserDataDeleter &onResponseUserDataDeleter
+    );
+
+private:
+    v8::Isolate* isolate_;
+    InspectorOnResponseCallback onResponse_;
+    InspectorUserDataDeleter onResponseUserDataDeleter_;
+};
+
+
+v8_inspector_channel_wrapper::v8_inspector_channel_wrapper(
+    v8::Isolate *isolate,
+    const InspectorOnResponseCallback &onResponse,
+    const InspectorUserDataDeleter &onResponseUserDataDeleter
+)
+  :
+    isolate_(isolate),
+    onResponse_(onResponse),
+    onResponseUserDataDeleter_(onResponseUserDataDeleter)
+{}
+
+void v8_inspector_channel_wrapper::sendResponse(int callId, std::unique_ptr<v8_inspector::StringBuffer> message) {
+    const std::string response = convertToString(isolate_, message->string());
+    if (onResponse_) {
+        onResponse_(response);
+    }
+}
+
+void v8_inspector_channel_wrapper::sendNotification(std::unique_ptr<v8_inspector::StringBuffer> message) {
+    const std::string notification = convertToString(isolate_, message->string());
+    if (onResponse_) {
+        onResponse_(notification);
+    }
+}
+
+void v8_inspector_channel_wrapper::flushProtocolNotifications() {
+    // flush protocol notification
+}
+
+void v8_inspector_channel_wrapper::setIsolate(v8::Isolate *isolate) {
+    isolate_ = isolate;
+}
+
+void v8_inspector_channel_wrapper::setOnResponseCallback(
+    const InspectorOnResponseCallback &callback,
+    const InspectorUserDataDeleter &deleter
+) {
+    if (onResponseUserDataDeleter_) {
+        onResponseUserDataDeleter_();
+    }
+
+    onResponse_ = callback;
+    onResponseUserDataDeleter_ = deleter;
+}
+
+class v8_inspector_client_wrapper final: public v8_inspector::V8InspectorClient {
+public:
+    explicit v8_inspector_client_wrapper(
+        v8::Platform *platform,
+        const v8::Local<v8::Context> &context,
+        const InspectorOnResponseCallback &onResponse = {},
+        const InspectorUserDataDeleter &onResponseUserDataDeleter = {},
+        const InspectorOnWaitFrontendMessageOnPauseCallback &onWaitFrontendMessageOnPause = {},
+        const InspectorUserDataDeleter &onWaitFrontendMessageOnPauseUserDataDeleter = {}
+    );
+
+    void setOnResponseCallback(
+        const InspectorOnResponseCallback &callback,
+        const InspectorUserDataDeleter &deleter
+    );
+    void setOnWaitFrontendMessageOnPauseCallback(
+        const InspectorOnWaitFrontendMessageOnPauseCallback &callback,
+        const InspectorUserDataDeleter &deleter
+    );
+
+    inline uint64_t getIsolateId() const noexcept {
+        const uint64_t *id_ptr = reinterpret_cast<uint64_t*>(isolate_->GetData(ISOLATE_ID_INDEX));
```

Comment:
The entire ID of the id check is not to access the isolate pointer on wrong id, so if by checking the id we access the isolate then we already lose. We should save the ID on the isolate when creating it, and then check against this local copy.